### ISSUE TITLE
UX Design Specs

### ISSUE THREAD

**Issue by coldkick** [meteor-gazelle/meteor-gazelle#98]
# Table of Contents
1. Introduction
2. Strategy Plane
   - Business Needs
   - User Needs
3. Scope Plane
4. Structure Plane
5. Skeleton Plane
6. Surface Plane
7. Accessibility
   - Colorblindness
   - Screen Readers
   - Dyslexia
#1. Introduction

Meteor allows us to take advantage of some powerful new isomorphic features to enhance a users experience while navigating a Meteor-Gazelle enabled site. Navigation can be _simplified_, and _responsive_. This issue should be used to discuss improvements to the UI.
#2. Strategy Plane

What are the goals of the product for our audience? What are our goals as developers? How does this influence our design and interaction with the site? The Strategy Plane is used to outline both the audience and business goals of the user experience.
## 2.1 Business Needs

TODO
## 2.2 User Needs

TODO
#3. Scope Plane

TODO
#4. Structure Plane

TODO
#5. Skeleton Plane

TODO
#6. Surface Plane
## Header Navigation

Top bar navigation contains the most commonly used links on the site. The current preview design has a very simple and effective top-bar navigation.

![](http://i.imgur.com/GZZ1ewK.png)
## Always-In-View Statistics

An important factor in a user's experience is their ability to be able to track their statistics in an always-in-view format. This is already applicable in the legacy gazelle, however it is lacking the isomorphic properties, requiring users to refresh the page and _waiting for 15 minutes_ for updates to roll through while up-to-date statistics can be found on the profile page. This breaks the homogeneous feeling of the site. Using mongo (or potentially Redis in the future) and isomorphic javascript through websockets or HTTP/2 these statistics can be kept real-time without users having to refresh the page allowing users to save bandwidth and improve their experience.

![](http://i.imgur.com/ABs7She.png)
## Contextual Menus

A major point of discussion, and the section that requires the most research. I'm using the term 'contextual' very broadly. I have two ideas for effective contextual menus.
1. Standard dropdown menus on click
   - Clicking the dropdown menu indicator on **Torrents** will open a standard dropdown menu with additional sections related to Torrent searching and statistics.
   - Clicking the dropdown menu indicator on **Profile Image** will open a standard dropdown menu with user settings, inbox, invites etc.
2. The searchbar menu will be reactive.
   - Clicking the dropdown menu indicator on **Torrents**  will change the contents to the left of the search bar to include additonal sections related to Torrent searching and statistics.
   - Clicking the dropdown menu indicator on **Profile Image**  will change the contents to the left of the search bar to include additional sections related to user settings, inbox, invites etc.
## Search

Legacy search is too granular. I for one give a big vote for the unified search box. I've modified the the design of the preview design. I for one, do not like the trend of hamburger buttons as they are not descriptive enough as to what the goal of the interaction is. Instead, I feel we should use FontAwesome's fa-filter. Clicking on the filter button will open a small menu where users can choose the section of the site to search. Upon clicking on their option the placeholder text in the search box would update using this template _Search {{section}}_

![](http://i.imgur.com/EfZyoav.png)
#7. Accessibility
## 7.1 Colorblindness

An issue often overlooked but while I personally love the mint green and auburn red design choice, users who are Red or Green deficient will have trouble distinguishing the two apart. It's not a major issue, but worth noting.
## 7.2 Screen Readers

ARIA Rules.
Least amount of front-end javascript as possible to ensure screen reader support.
## 7.3 Dyslexia

@ajax-wcd @cosm @sneakysneakysneaky @Malvar

**Comment by sneakysneakysneaky** [comment]
@coldkick, thank you for doing this. Very well put together.

My thoughts:
- Always-in-view statistics will be reactive. The underlying technology (Meteor subscriptions vs websockets vs HTTP2, etc) should be abstracted from a UX discussion such as this.
- It is not obvious from the current image, but the hamburger button will toggle between unified & granular search, like so:
  ![01d-search-expanded](https://cloud.githubusercontent.com/assets/13206949/9026222/8ede15e0-38f5-11e5-9d1e-3bf67ef60725.jpg)
  - That said, I support replacing the hamburger icon with the filter icon. 
- I'm afraid i cannot support what you call "reactive menus."
  - As a user, these kinds of menus annoy me. They always end up being a moving target, & i never know where my desired element is going to appear.
  - This also presents a challenge on how they will behave on smaller screen sizes. See the next point
- We need to think about what the experience will be for small screen sizes & touch screens. The current mockup (and codebase, for that matter) appears quite wide. How can we optimize site navigation on small screens?
- Also remember that users can apply different themes. The UX needs to both be awesome & intuitive, yet simple enough so the stylesheet can be swapped out at runtime. It shall be quite the balancing act.

**Comment by coldkick** [comment]
Thanks for the response, this is still very much a WIP.

The granular search with mobile devices will be an issue as well, will it become a grid, a vertical bar?
Perhaps on smaller devices it would make more sense to work as I explained, with a dropdown toggle between torrents, artists, labels, requests, forums, log, and users.

On reactive menus: After I wrote this I did some more looking around and found your header spec: https://github.com/meteor-gazelle/meteor-gazelle/blob/master/doc/specs/header.md
which explained to me the goals of the toggles on the header navigation. I will edit the comment to reflect this later.

Were collages moved out of being a secondary link under Torrents? The subcategories under Help need to be fixed as well. Other than that, the spec felt very complete.

**Comment by borghunter-wcd** [comment]
Some of my own thoughts, and things to think about further...

#### Always-In-View Statistics
1. It would be useful if these stats were only shown if pertinent. For example: 
   - A user ranked high enough could have infinite invites, so the "Invites: X" stat would be completely useless.
   - The "required" stat's importance changes. A users with a ratio of "4" and a required of "0.6", won't find it useful. However, someone sitting at "0.7" would. Also, a user may be more interested in the required ratio to keep them at their current rank, rather than the ratio that would keep them from being disabled or whatever.

#### Search
1. Hamburger icons are (at least for me) meant to convey a more general navigation menu. Would be useful on mobile for the primary navigation menu, but that's about it. the fa-filter icon is great.
2. For mobile, and even desktop, perhaps something like this when searching a specific section... Then when using the universal search, there would be sections, probably prioritizing artists with a "Browse all X" every 5 results or so.

![](http://2.bp.blogspot.com/-nX3JOliiRUk/VN38P7Q00AI/AAAAAAACAgU/AA1gxc6PXI0/s1600/flight-search-help-1.png)

#### Colorblindness

Eff yes. I've worked a bit in web accessibility, but am no expert. If there's someone who can generally keep on top of that for every section, that would be great. Having it part of our coding and design standards would be great.

**Comment by chriscubellis** [comment]
Hey– @cosm here. 

I haven't been able to be too involved other than the designs shown due to time constraints, but the write-up above all looks really great, thanks for doing that all.

Here is my working Dropbox folder of current .psd's & jpegs (really just index.psd / artist.psd at the moment): https://www.dropbox.com/sh/xmfhqqbvad6pgyc/AADpfJUPH4YqGpIv272o0ryea?dl=0

The folder contains jpegs of all of the index states & artist page states. The artist page has a toggle between full list view and album view, album view hovers basically activate list view elements one at a time in a sort-of google images-esque slidedown. 

For the collages, I was imaging that the top 4 or 8 would be shown, with the rest (if applicable) behind a view all button.

**Direct links**

[Index.psd](https://www.dropbox.com/s/4mh74zt1vi528zh/00-index.psd?dl=0)
[Artist.psd](https://www.dropbox.com/s/7x9ohej66qtifae/02-artist.psd?dl=0)

Apop is the other designer on this, not sure of his current progress (know he was dealing with some illness).

When it comes down to it, I don't have a terrible amount of time to dedicate to this and you guys know the site better than me. My main goal helping out here is to lay out a guideline of modules and general design layouts that you guys can tweak as needed in development.

Quick responses to some of the stuff higher up on this page–
- I don't at all support replacing the hamburger icon with whatever that "fa-filter" icon is. I don't consider myself a huge fan of hamburger menus in general, but this is a good use case of them (revealing additional options). It's a learned behavior that is extremely rampant across navigation paradigms that essentially every internet-user has some inclination as to what it does. I do this for a living and have...never seen that icon before. So probably no one else will really know what it does either.
- Real time stat updating sounds like an amazing improvement over the current legacy stats
- Colorblindness– really good call. I don't have much experience designing for accessibility, but I imagine that there'd have to be a theme that has a color set more applicable to the color blind.

**Comment by sneakysneakysneaky** [comment]
Thanks for the PSDs, @chriscubellis. Stylesheet development can go on!

I'll go with the consensus with the hamburger vs filter icon. Good points have been raised for using either icon.

I do have some experience developing accessible stylesheets. I am not too worried. The red & green colors do have similar shades, but the contrast between them & the surrounding elements is good, & the colors don't seem to have any semantic value. The only thing i see right now is to change more than the color for the navigation's hover style.

**Comment by erl7902** [comment]
Hey so I've just kind of been lurking so far, but I actually just finished an internship developing software for visually impaired kids (from colorblind to completely blind). If you need any input/testing let me know and I'll be happy to help out

**Comment by salas-wcd** [comment]
I think not only colorblindness but there is also some people suffering from dyslexia. For thoose,  very bright page with almost no border is hard to follow. And for those,  clear separation between elements (like material design) is a reall plus.  It may be great to support a official bright,  an official dark and an official accessible theme (right font,  right color right contrast).

**Comment by salas-wcd** [comment]
Else on what,  on the laboratory there was a project of designing the best theme ever. The aithor was not a web designer but there is really great mockups and comments that can inspire us.

**Comment by sneakysneakysneaky** [comment]
There will be an official light & dark stylesheet.

**Comment by coldkick** [comment]
I'll edit the issue to reflect a larger scope Accessibility section.
In a way I feel that accessibility should get a separate issue. Screen readers need particular ARIA tags for visual impairment. Menus should not be in unordered lists but instead be in a nav with a tags and a nav a { display: block; } rule

**Comment by salas-wcd** [comment]
About hamburger icon : http://deep.design/the-hamburger-menu/

**Comment by coldkick** [comment]
Thoughts on this format for search etc:

_Desktop:_ Default atomic search. When page width becomes less than X default becomes the unified search. All icons that will be used can be hovered over for a tooltip explaining their purpose. Similar to the current tooltips

_Mobile:_ Default unified search. A javascript 'helper' button, using perhaps a question mark as a symbol that will automatically open all tooltips giving similar functionality as its desktop counterpart.
This both clarifies the interface to new users and can be used as a quick reference in the future, or as a guide to any new features added to the project.

**Comment by chriscubellis** [comment]
@salas-wcd There are infinite articles both in favor of and against the hamburger icon used throughout the web, but if you're gunna link to someone else's argument without elaboration, it should at least be applicable to the use case being discussed.

A super majority of that article is referring to using the hamburger icon as a replacement to having a primary nav...immediately not the scenario here. Interestingly enough, using the hamburger menu here actually _allows_ us to show all of our secondary nav (the primary nav is always shown regardless). 

In this case, with search expanded into its separate fields (e.g., hamburgerless-mode) all of the user links get folded into a dropdown. In hamburger mode/unified search, all of those user links are immediately accessible. 

But back to the article, at the very bottom there's an addendum listing out some exceptions to the "rule". All three of those exceptions are applicable to this use case.

Bolded the points that argue in favor of this use-case of a hamburger icon.

> For example, a repetitive-use application with trained users (e.g., a medical application used by nurses throughout the course of the day) may be a valid scenario for using such an interface. **“Power user” systems** don’t always need to make things intuitive for new users, and sacrifices can and should sometimes be made in order **to give additional benefits—such as information density—**to experienced users.
> 
> Another example might be (as seen in the Facebook example above) **truly secondary navigation items. Once the most important elements have been explicitly listed for the user**, you can scoop up all the miscellaneous additional stuff that nobody’s going to look for anyway and throw it under a low-priority hamburger menu.
> 
> Universally declaring hamburger menus to be totally terrible is as bad as declaring them to be always awesome. **Rules are no replacement for thinking. Consider the user, do the math, and figure out what actually makes the most sense.**

Hamburger icons can be argued to death, but this actually is an extremely valid use case for a hamburger icon. 

(And if it's a matter of hamburger icon versus other icon that performs the same action....hamburger menu is generally the better option because of user awareness, everyone who uses the internet knows that clicking a hamburger icon is going to reveal more options.)

**Comment by salas-wcd** [comment]
Sorry, was send on my mobile. For me hamburger icon alone is all but clear.
A directional like this seems more clear like following.

![image](https://cloud.githubusercontent.com/assets/13600559/9187687/8dbb39a6-3fd3-11e5-9201-05488c05180a.png)

And for search page in order to display an advanced search I prefer a button with text, so much easier to  

![image](https://cloud.githubusercontent.com/assets/13600559/9187662/55bbc7aa-3fd3-11e5-8a5a-9e14c45c1c11.png)

But yes it is personal point of view. I just found this hn article pretty reliant to the UX spec and also pretty nice to have good pratice for mobile dev.

**Comment by coldkick** [comment]
That is what I envisioned for the unified search as well. I feel that after a certain x min width, the site should switch to the filtered/unified mode with a drop down choice of category

+1

**Comment by chriscubellis** [comment]
I totally agree that that sort of search functionality would work way better & I have no qualms using it, but the reason I had that hamburger toggle is because the crowd in the what forum thread seemed split between people who wanted legacy search all of the time and people that wanted unified search...so the hamburger expand into legacy seemed to satiate both of those options.

**Comment by ghost** [comment]
@coldkick @chriscubellis We can't remove the option for legacy search. There was a lot of discussion about it internally and there's a larger than expected camp of people who want legacy search. I don't understand why that is but as it stands now we have to support legacy search in one form or another.

**Comment by coldkick** [comment]
@ajax-wcd I'm vouching for legacy search to remain for devices over an arbitrary width such as say 1000 px. Any smaller than that and the search bars would cause the window to scroll or they would become too small to be of use. I agree that it should remain, this is more for smaller devices, phablets, mobile, and the such.

**Comment by ghost** [comment]
@coldkick Yep.
